citest/reporting/html_document_manager.py

```diff
@@ -22,10 +22,112 @@
 import cgi
 
 
+class MyTag(object):
+  """Denotes an HTML tag and its content.
+
+  This interface is a minimal subset of BeautifulSoup that we use on tags.
+  The class is a more efficient implementation that runs in 25% of the time
+  (but still about 2.5x longer than hand-constructing HTML).
+  """
+  # pylint: disable=too-few-public-methods
+
+  def __init__(self, name, **kwargs):
+    """Construct a tag
+
+    Args:
+      name: [string] The tag name
+      kwargs: [kwargs] Optional tag key/value attributes
+    """
+    self.__name = name
+    self.__attrs = dict(kwargs)
+    self.__parts = []
+    self.string = ''
+
+  def __setitem__(self, key, value):
+    """Add additional key/value attributes."""
+    self.__attrs[key] = value
+
+  def append(self, part):
+    """Add additional contents (children) to a tag.
+
+    Args:
+      part: [MyTag/String/DocFragment] The thing to add.
+    """
+    self.__parts.append(part)
+
+  def __str__(self):
+    attributes = [' {key}="{value}"'.format(key=key, value=value)
+                  for key, value in self.__attrs.items()]
+    return '<{name}{attributes}>{text}{parts}</{name}>'.format(
+        name=self.__name,
+        attributes=''.join(attributes),
+        text=self.string,
+        parts=''.join([str(part) for part in self.__parts]))
+
+
+class MyDocFragment(object):
+  """Denotes a block of HTML text.
+
+  This interface is a minimal subset of BeautifulSoup that we use.
+  Note that unlike BeautifulSoup, we wont parse raw HTML we are given.
+  Therefore we wont be able to navigate or edit the internal structure.
+  However, since we are constructing the document, we shouldnt need to.
+
+  The class is a more efficient implementation that runs in 25% of the time
+  (but still about 2.5x longer than hand-constructing HTML).
+  """
+  def __init__(self, html):
+    """Construct fragment from initial HTML.
+
+    Args:
+      html: [string] Escaped HTML text that starts the fragment.
+    """
+    self.__parts = [html]
+
+  def append(self, part):
+    """Add a block to the end of the current fragment.
+
+    Args:
+      part: [any]  Part shuld be a Tag/String/DocFragment
+    """
+    self.__parts.append(part)
+
+  def __str__(self):
+    return ''.join([str(part) for part in self.__parts])
+
+  def new_tag(self, name, **kwargs):
```

Comment:
Alright, makes sense if we do need the HTML parsing/searching. I was thinking from a perspective where we are just building HTML, possibly using a different library to do the same thing.